blockchain/v2/reactor.go

```diff
@@ -1,118 +1,521 @@
 	go r.scheduler.start()
 	go r.processor.start()
 	go r.demux()
+	return nil
+}
 
-	<-r.scheduler.ready()
-	<-r.processor.ready()
+// reactor generated ticker events:
+// ticker for cleaning peers
+type rTryPrunePeer struct {
+	priorityHigh
+	time time.Time
+}
 
-	go func() {
-		for t := range r.ticker.C {
-			r.events <- timeCheck{time: t}
-		}
-	}()
+func (e rTryPrunePeer) String() string {
+	return fmt.Sprintf(": %v", e.time)
+}
+
+// ticker event for scheduling block requests
+type rTrySchedule struct {
+	priorityHigh
+	time time.Time
+}
+
+func (e rTrySchedule) String() string {
+	return fmt.Sprintf(": %v", e.time)
+}
+
+// ticker for block processing
+type rProcessBlock struct {
+	priorityNormal
+}
+
+// reactor generated events based on blockchain related messages from peers:
+// blockResponse message received from a peer
+type bcBlockResponse struct {
+	priorityNormal
+	time   time.Time
+	peerID p2p.ID
+	size   int64
+	block  *types.Block
+}
+
+// blockNoResponse message received from a peer
+type bcNoBlockResponse struct {
+	priorityNormal
+	time   time.Time
+	peerID p2p.ID
+	height int64
+}
+
+// statusResponse message received from a peer
+type bcStatusResponse struct {
+	priorityNormal
+	time   time.Time
+	peerID p2p.ID
+	height int64
+}
+
+// new peer is connected
+type bcAddNewPeer struct {
+	priorityNormal
+	peerID p2p.ID
+}
+
+// existing peer is removed
+type bcRemovePeer struct {
+	priorityHigh
+	peerID p2p.ID
+	reason interface{}
 }
 
-// XXX: How to make this deterministic?
-// XXX: Would it be possible here to provide some kind of type safety for the types
-// of events that each routine can produce and consume?
-func (r *Reactor) demux() {
+func (r *BlockchainReactor) demux() {
+	var lastRate = 0.0
+	var lastHundred = time.Now()
+
+	var (
+		processBlockFreq = 20 * time.Millisecond
+		doProcessBlockCh = make(chan struct{}, 1)
+		doProcessBlockTk = time.NewTicker(processBlockFreq)
+
+		prunePeerFreq = 1 * time.Second
+		doPrunePeerCh = make(chan struct{}, 1)
+		doPrunePeerTk = time.NewTicker(prunePeerFreq)
+
+		scheduleFreq = 20 * time.Millisecond
+		doScheduleCh = make(chan struct{}, 1)
+		doScheduleTk = time.NewTicker(scheduleFreq)
+
+		statusFreq = 10 * time.Second
+		doStatusCh = make(chan struct{}, 1)
+		doStatusTk = time.NewTicker(statusFreq)
+	)
+
+	// XXX: Extract timers to make testing atemporal
 	for {
 		select {
+		// Pacers: send at most per frequency but don't saturate
+		case <-doProcessBlockTk.C:
+			select {
+			case doProcessBlockCh <- struct{}{}:
+			default:
+			}
+		case <-doPrunePeerTk.C:
+			select {
+			case doPrunePeerCh <- struct{}{}:
+			default:
+			}
+		case <-doScheduleTk.C:
+			select {
+			case doScheduleCh <- struct{}{}:
+			default:
+			}
+		case <-doStatusTk.C:
+			select {
+			case doStatusCh <- struct{}{}:
+			default:
+			}
+
+		// Tickers: perform tasks periodically
+		case <-doScheduleCh:
+			r.scheduler.send(rTrySchedule{time: time.Now()})
```

Comment:
can't we merge this with above cases? i.e.

```go
case <-doScheduleTk.C:
  r.scheduler.send(rTrySchedule{time: time.Now()})
```

call scheduler directly?

Reply:
Indeed, will address in a follow up PR #4410.